### PR TITLE
Community sub button

### PR DESCRIPTION
**Pull request by krawieck**
* safe from button mashing
* spinner delayed by 500 ms
* shows snack bar on error
* it works

also replaced constructors with `useMemoFuture()` cuz of no access to store without `context`

Screen Recording 2020-09-15 at 16 12 15

### CODE REVIEW

File: lib/util/token_or_null.dart [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+/// have any user logged in
+String tokenOrNull(BuildContext context, String instanceUrl) {
+  if (context
+      // ignore: avoid_types_on_closure_parameters
+      .select((AccountsStore store) => store.isAnonymousFor(instanceUrl))) {
+    return null;
+  }
+  return context
+      // ignore: avoid_types_on_closure_parameters
+      .select((AccountsStore store) => store.defaultTokenFor(instanceUrl).raw);
```
Comment (shilangyu) [inline lib/util/token_or_null.dart:17]: use `context.select<TYPE>(...)`

File: lib/util/token_or_null.dart [hunk 2]
```diff
@@ -0,0 +1,18 @@
+import 'package:flutter/material.dart';
+import 'package:provider/provider.dart';
+
+import '../stores/accounts_store.dart';
+
+/// returns either raw token for default user of `instanceUrl`
+/// or null if this instance is not added or doesn't
+/// have any user logged in
+String tokenOrNull(BuildContext context, String instanceUrl) {
```
Comment (shilangyu) [inline lib/util/token_or_null.dart:9]: does not handle the case where someone does not have this instance added

Please modify `AccountsStore.isAnonymousFor` in a way where it checks if the instance is added (which would then mean it is anonymous)

File: lib/pages/community.dart [hunk 3]
```diff
@@ -7,47 +9,59 @@ import 'package:intl/intl.dart';
...
+        communityId = _community.id,
+        communityName = _community.name;
 
   @override
   Widget build(BuildContext context) {
     final theme = Theme.of(context);
-    var fullCommunitySnap = useFuture(_fullCommunityFuture);
+    var fullCommunitySnap = useMemoFuture(() {
+      final auth = tokenOrNull(context, instanceUrl);
+      if (communityId != null && instanceUrl != null) {
```
Comment (shilangyu) [inline lib/pages/community.dart:53]: `instanceUrl` is required, why would it be null?

File: lib/pages/community.dart [hunk 4]
```diff
@@ -541,3 +522,91 @@ class _Divider extends StatelessWidget {
...
+    final colorOnTopOfAccent = textColorBasedOnBackground(theme.accentColor);
+    final subscribed = community.subscribed ?? false;
+    final token = tokenOrNull(context, community.instanceUrl);
+
+    final isSubbed = useState(subscribed);
+    final showSpinner = useState(false);
+    final isPending = useState(false);
+
+    subscribe() async {
+      if (token == null) {
```
Comment (shilangyu) [inline lib/pages/community.dart:543]: use dedicated methods for checking anonymity (`AccountsStore.isAnonymousFor`)

File: lib/pages/community.dart [hunk 5]
```diff
@@ -541,3 +522,91 @@ class _Divider extends StatelessWidget {
...
+  @override
+  Widget build(BuildContext context) {
+    final theme = Theme.of(context);
+    final colorOnTopOfAccent = textColorBasedOnBackground(theme.accentColor);
+    final subscribed = community.subscribed ?? false;
+    final token = tokenOrNull(context, community.instanceUrl);
+
+    final isSubbed = useState(subscribed);
+    final showSpinner = useState(false);
+    final isPending = useState(false);
```
Comment (shilangyu) [inline lib/pages/community.dart:540]: extract into a reusable hook

File: lib/pages/community.dart [hunk 6]
```diff
@@ -541,3 +522,91 @@ class _Divider extends StatelessWidget {
...
+        Scaffold.of(context).showSnackBar(
+            SnackBar(content: Text("can't sub when you're not logged in")));
+        return;
+      }
+
+      isPending.value = true;
+      var spinnerTimer =
+          Timer(Duration(milliseconds: 500), () => showSpinner.value = true);
+
+      final api = LemmyApi(community.instanceUrl).v1;
```
Comment (shilangyu) [inline lib/pages/community.dart:553]: This variable is created and used once, maybe it was intentional

File: lib/pages/community.dart [hunk 7]
```diff
@@ -541,3 +522,91 @@ class _Divider extends StatelessWidget {
...
+        // ignore: avoid_catches_without_on_clauses
+      } catch (e) {
+        Scaffold.of(context).showSnackBar(SnackBar(
+            content: Row(
+          children: [
+            Icon(Icons.warning),
+            SizedBox(width: 10),
+            Text("couldn't ${isSubbed.value ? 'un' : ''}sub :<"),
+          ],
+        )));
```
Comment (shilangyu) [inline lib/pages/community.dart:567]: please add one trailing comma, this formatting makes me anxious

**Review by shilangyu — CHANGES_REQUESTED** [review]

File: lib/pages/community.dart (hunk 4 above)
Comment (krawieck) [reply lib/pages/community.dart:543]: no. i do this checking in `tokenOrNull` to reduce boilerplate code

File: lib/pages/community.dart (hunk 7 above)
Comment (krawieck) [reply lib/pages/community.dart:567]: lmao

**Pull request merged**